server/odc-service/src/main/java/com/oceanbase/odc/service/monitor/MeterHolder.java

```diff
@@ -0,0 +1,363 @@
+            this.registry = registry;
+            this.monitorProperties = monitorProperties;
+        }
+
+        public void increment(MeterKey meterKey, Integer defaultValue) {
+            GaugeHolder.GaugeWrapper wrapper = GAUGE_MAP.get(meterKey);
+            if (wrapper == null) {
+                if (GAUGE_MAP.size() >= monitorProperties.getMeter().getMaxGaugeMeterNumber()) {
+                    log.error("Too many gauges to register, meterKey={}", meterKey);
+                    return;
+                }
+                GAUGE_MAP.computeIfAbsent(meterKey, (key) -> new GaugeHolder.GaugeWrapper(null, null));
+            }
+            MeterKey lockKey = getLockKey(meterKey);
+            synchronized (lockKey) {
+                wrapper = GAUGE_MAP.get(meterKey);
+                if (wrapper.value == null) {
+                    Gauge gauge = registerGauge(meterKey);
+                    wrapper.setValue(defaultValue);
+                    wrapper.setGauge(gauge);
+                }
+                GAUGE_MAP.get(lockKey).increment(1);
+            }
+        }
+
+        public void decrement(MeterKey meterKey, Integer defaultValue) {
+            GaugeHolder.GaugeWrapper wrapper = GAUGE_MAP.get(meterKey);
+            if (wrapper == null) {
+                if (GAUGE_MAP.size() >= monitorProperties.getMeter().getMaxGaugeMeterNumber()) {
+                    log.error("Too many gauges to register, meterKey={}", meterKey);
+                    return;
+                }
+                GAUGE_MAP.computeIfAbsent(meterKey, (key) -> new GaugeHolder.GaugeWrapper(null, null));
+            }
+            MeterKey lockKey = getLockKey(meterKey);
+            synchronized (lockKey) {
+                wrapper = GAUGE_MAP.get(meterKey);
+                if (wrapper.value == null) {
+                    registerGauge(meterKey);
+                    wrapper.setValue(defaultValue);
+                }
+                GAUGE_MAP.get(lockKey).decrement(1);
+            }
+        }
+
+        private MeterKey getLockKey(MeterKey meterKey) {
+            return GAUGE_MAP.keySet().stream().filter(meterKey::equals).findFirst()
+                    .orElseThrow(IllegalAccessError::new);
+        }
+
+        private Gauge registerGauge(MeterKey meterKey) {
+            Iterable<Tag> tags = meterKey.getTags();
+            Gauge.Builder<Supplier<Number>> builder = Gauge.builder(meterKey.getMeterName().getMeterName(),
+                    () -> GAUGE_MAP.get(meterKey).getValue())
+                    .description(meterKey.getMeterName().getDescription());
+            for (Tag tag : tags) {
+                builder.tag(tag.getKey(), tag.getValue());
+            }
+            return builder.register(registry);
+        }
+
+        public Integer getSize() {
+            return GAUGE_MAP.size();
+        }
+
+        @Data
+        @AllArgsConstructor
+        static class GaugeWrapper {
+            Integer value;
```

Comment:
gauge should be thread safe, use thread safe structure like atomic, by the way, is Integer big enough for metric?

Reply:
removed